requirements/updater.sh

```diff
@@ -0,0 +1,78 @@
+#!/bin/sh
+set -ue
+
+export LC_ALL=C
+requirements_in="$(readlink -f ./requirements.in)"
+requirements_ansible_in="$(readlink -f ./requirements_ansible.in)"
+requirements="$(readlink -f ./requirements.txt)"
+requirements_ansible="$(readlink -f ./requirements_ansible.txt)"
+pip_compile="pip-compile --quiet -r --allow-unsafe"
+
+_cleanup() {
+  cd /
+  test "${KEEP_TMP:-0}" = 1 || rm -rf "${_tmp}"
+}
+
+install_deps() {
+  pip install pip --upgrade
+  pip install pip-tools
+}
+
+generate_requirements_v3() {
+  venv="./venv3"
+  python3 -m venv "${venv}"
+  # shellcheck disable=SC1090
+  . "${venv}/bin/activate"
+
+  install_deps
+
+  ${pip_compile} --output-file requirements.txt "${requirements_in}"
+  ${pip_compile} --output-file requirements_ansible_py3.txt "${requirements_ansible_in}"
+}
+
+generate_requirements_v2() {
+  venv="./venv2"
+  virtualenv -p python2 "${venv}"
+  # shellcheck disable=SC1090
+  PS1="" . "${venv}/bin/activate"
+
+  install_deps
+
+  ${pip_compile} --output-file requirements_ansible.txt "${requirements_ansible_in}"
+}
+
+generate_patch() {
+  a="requirements_ansible_py3.txt"
+  b="requirements_ansible.txt"
+
+  # most elegant/quick solution I could come up for now
+  out="$(diff --ignore-matching-lines='^#' --unified "${a}" "${b}" | \
+    awk '{ if (/^+\w/){ $2="; python_version < \"3\" #"; print;} else print; }' | \
+    sed 's/ ;/;/g')"
+  test -n "${out}"
+  echo "${out}"
+}
+
+main() {
+  _tmp="$(mktemp -d --suffix .awx-requirements XXXX -p /tmp)"
+
+  if [ "$1" = "upgrade" ]; then
+      pip_compile="${pip_compile} --upgrade"
+  fi
+
+  cd "${_tmp}"
+  trap _cleanup INT TERM EXIT
+  generate_requirements_v3
+  generate_requirements_v2
+  cp -vf requirements.txt "${requirements}"
+  sed -i 's/^docutils.*//g' "${requirements}"
```

Comment:
Is there anyone who knows why we should remove the docutils?

Also should this be removed for `requirements.txt` and `requirements_ansible.txt`?

Reply:
> Is there anyone who knows why we should remove the docutils?

The reason is because `requirements_git.txt` is installed along side of the `requirements.txt`. Hypothetically the same would be true for `requirements_ansible_git.txt`, but that no longer has any entries, so it can be ignored.

Even the `requirements_git.txt` wouldn't be a problem, if not for the fact that some other library just happened to list docutils as its own dependency.

Reply:
I suppose you are referring to this?
https://github.com/ansible/awx/blob/38aedcdd4885bb50eb7578b1c0012ea91f9b6c18/Makefile#L174-L182

I'm confused why this got implemented/seperated in this way. I can only find this one place where the requirement files are getting installed - why are they seperated into so many files?

Reply:
"reasons" - the way we do some of the things here is because of nuanced and non-obvious constraints imposed by our downstream build system. There are a million ways to handle python packaging and none of them really work all that well - what we have here is years of effort towards reproducible and stable builds.

Reply:
Just out of curiosity (also for possible future contributions): Where can I see downstream build system source? Mostly interested in packaging and how the Docker images are built (awx_task and awx_web). On Dockerhub you can usually link the Dockerfile, but not in this case.

Reply:
Ah, found it myself https://github.com/ansible/awx/tree/devel/installer/roles/image_build